### PR TITLE
Implement Encrypted Streams

### PR DESCRIPTION
**Pull request by Kleidukos**
## TODO

- [x] `Multipart` continuation function
- [x] Usage example at the top
- [ ] push / pull with additional data

image

### CODE REVIEW

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:119]: We do this in a bunch of places. Might be worth writing a helper `newtype` for DerivingVia for these.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+-- If the message is not strictly longer than the size of the authentication tag,
+-- the function will fail.
+--
+-- @since 0.0.1.0
+cipherTextFromByteString :: StrictByteString -> Maybe CipherText
+cipherTextFromByteString bytestring =
+  let (foreignPtr, bsLength) = BS.toForeignPtr0 bytestring
+   in if bsLength > fromIntegral @CSize @Int cryptoSecretStreamXChaCha20Poly1305ABytes
+        then Just $ CipherText (Foreign.castForeignPtr @Word8 @CUChar foreignPtr) (fromIntegral @Int @CSize bsLength)
+        else Nothing
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:160]: I'm not sure this is completely safe: we copy a pointer to some foreign data, but not the data itself, which means that when the `ByteString` drops out of scope, our pointer now dangles. An 'owning copy' would be much better.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+
+-- | Convert a 'CipherText' to a binary 'StrictByteString'.
+--
+-- @since 0.0.1.0
+cipherTextToBinary :: CipherText -> StrictByteString
+cipherTextToBinary (CipherText fPtr size) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr fPtr)
+    (fromIntegral size)
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:169]: Similar to `cipherTextFromByteString`.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+--
+-- If the bytestring is not of the appropriate length, the function will fail.
+--
+-- @since 0.0.1.0
+headerFromByteString :: StrictByteString -> Maybe Header
+headerFromByteString bytestring =
+  let (foreignPtr, bsLength) = BS.toForeignPtr0 bytestring
+   in if bsLength == fromIntegral @CSize @Int cryptoSecretStreamXChaCha20Poly1305HeaderBytes
+        then Just $ Header (Foreign.castForeignPtr @Word8 @CUChar foreignPtr)
+        else Nothing
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:255]: Here as well.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Enum
+      -- ^ @since 0.0.1.0
+    , Bounded
+      -- ^ @since 0.0.1.0
+    )
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:283]: Needs a `stock`.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+  , mAdditionalData :: Maybe StrictByteString
+  }
+  deriving
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    )
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:314]: Also here.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+  | InputStreamInitError
+  | InputStreamPushError
+  deriving
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    )
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:336]: Here also.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+  Foreign.withForeignPtr headerForeignPtr $ \headerPtr ->
+    Foreign.withForeignPtr secretKeyForeignPtr $ \keyPtr -> do
+      result <-
+        cryptoSecretStreamXChaCha20Poly1305InitPush
+          statePtr
+          headerPtr
+          keyPtr
+      case result of
+        0 -> pure $ Right (Header headerForeignPtr, SecretKey secretKeyForeignPtr)
+        _ -> pure $ Left InputStreamInitError
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:366]: This leaks memory, unless `mallocForeignPtrBytes` automagically installs a finalizer.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+              Foreign.nullPtr
+              (Foreign.castPtr @CChar @CUChar cString)
+              (fromIntegral @Int @CULLong cStringLen)
+              additionalDataPointer
+              additionalDataLength
+              (streamTagToConstant tag)
+          case result of
+            0 -> do
+              Foreign.withForeignPtr cipherTextFPtr $ \cipherTextHomePtr ->
+                memcpy cipherTextBuffer cipherTextHomePtr (fromIntegral cipherTextLength)
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:409]: This is a bit inefficient: you allocate a temporary buffer, only to then copy it immediately on success.

**Review by kozross — CHANGES_REQUESTED** [review]

File: sel/src/Sel/SecretKey/EncryptedStream.hs (hunk 2 above)
Comment (Kleidukos) [reply sel/src/Sel/SecretKey/EncryptedStream.hs:160]: Something like this?

```haskell
      cipherForeignPtr <- Foreign.mallocPlainForeignPtrBytes (BS.length bytestring)
      Foreign.withForeignPtr cipherForeignPtr $ \cipherPtr ->
        BS.unsafeUseAsCStringLen bytestring $ \(cString, cStringLen) -> do
          memcpy (Foreign.castPtr cipherPtr) cString (fromIntegral cStringLen)
          pure $
            Just $
              CipherText
                cipherForeignPtr
                (fromIntegral @Int @CSize cStringLen)
```

File: sel/src/Sel/SecretKey/EncryptedStream.hs (hunk 8 above)
Comment (Kleidukos) [reply sel/src/Sel/SecretKey/EncryptedStream.hs:366]: Actually yes! https://hackage.haskell.org/package/base-4.18.0.0/docs/Foreign-ForeignPtr.html#v:mallocForeignPtr

File: sel/src/Sel/SecretKey/EncryptedStream.hs (hunk 2 above)
Comment (kozross) [reply sel/src/Sel/SecretKey/EncryptedStream.hs]: Does `mallocPlainForeignPtrBytes` give us a finalizer?

File: sel/src/Sel/SecretKey/EncryptedStream.hs (hunk 8 above)
Comment (kozross) [reply sel/src/Sel/SecretKey/EncryptedStream.hs:375]: Huh, good to know.

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+-- @since 0.0.1.0
+initPullStream
+  :: Multipart s
+  -- ^ Cryptographic state.
+  -> Header
+  -- ^ Header received from the peer, used to ensure the ciphertexts will be valid.
+  -> SecretKey
+  -- ^ Secret key from the peer, used to decrypt the ciphertexts.
+  -> IO (Either EncryptedStreamError ())
+  -- ^ Returns 'True' on success, 'False' if the header is invalid
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:519]: Was this meant to return an `IO (Either EncryptedStreamError Bool)`?

File: sel/src/Sel/SecretKey/EncryptedStream.hs [hunk 11]
```diff
@@ -0,0 +1,650 @@
...
+-- Applications will typically call this function in a loop,
+-- until a message with the 'Final' tag is found.
+--
+-- If the tag cannot be decoded from the payload, then it is not returned in the `StreamResult`.
+--
+-- @since 0.0.1.0
+pullFromStream
+  :: Multipart s
+  -> CipherText
+  -> Maybe Word
```
Comment (kozross) [inline sel/src/Sel/SecretKey/EncryptedStream.hs:542]: What's this argument meant to do?

**Review by kozross — CHANGES_REQUESTED** [review]

File: sel/src/Sel/SecretKey/EncryptedStream.hs (hunk 10 above)
Comment (Kleidukos) [reply sel/src/Sel/SecretKey/EncryptedStream.hs:519]: Actually the doc is wrong, I'm experimenting with more expressive error types.

**Comment by Kleidukos** [comment]
Closing in favour of something based on a pre-existing streaming library

**Pull request closed without merging**